fix: display tool call/return content in messages, recover response from false-failed runs

## Context

While debugging a production Draper agent run, two lettactl gaps surfaced:

1. `get messages --system` shows tool_call/tool_return message types but hides their content
2. Letta server occasionally marks runs as `failed` even when all steps succeeded and the assistant_message was produced — lettactl treats these as hard failures with no recovery

Both were discovered on a real run (`run-ccdc7afc-2b90-4024-8e4b-5458fc831dc8`) where Draper completed 14 steps successfully, produced the correct IMAGE/VIDEO/CONCEPT output, but the run was marked `status: 'failed'`, `stop_reason: 'error'`, `metadata.error: ""` (empty string). Context window was at 60,111 / 64,000 tokens on the last step — likely Letta errored during post-run bookkeeping (memory consolidation / context compaction) after the agent work was done.

---

## Fix 1: Display tool_call and tool_return content in `get messages --system`

### Problem

When using `lettactl get messages <agent> --system`, tool messages render as:

```
10:58:10 PM [tool_call_message]
  [tool_call_message]

10:58:10 PM [tool_return_message]
  [tool_return_message]
```

The content is there in the message object (`tool_call.name`, `tool_call.arguments`, `tool_return`, `status`) — it just isn't extracted.

Hitting the Letta API directly (`GET /v1/runs/{runId}/messages`) returns full content:

```json
{
  "message_type": "tool_call_message",
  "tool_call": {
    "name": "start_image_generation",
    "arguments": "{\"company_id\": \"...\", \"prompt\": \"Wide-angle view of a premium airplane cabin...\"}",
    "tool_call_id": "toolu_bdrk_016Z3JqoiH3QaqY3ty4fukc2"
  },
  "tool_calls": [...]
}

{
  "message_type": "tool_return_message",
  "tool_return": "Image generation started. Job ID: aaa000b6-0fce-4dfa-a0ac-8905cb41b270",
  "status": "success",
  "tool_returns": [...]
}
```

### Root Cause

`getMessageContent()` in `src/commands/messages/utils.ts` (line 4-19) only checks `.text`, `.content`, and `.message` — it has no handling for `tool_call_message` or `tool_return_message` structures.

The display function in `src/lib/ux/display/messages.ts` (line 25) falls back to `[${msg.role}]` when content is null.

### Proposed Fix

Update `getMessageContent()` to handle these message types:

```typescript
export function getMessageContent(message: any): string | null {
  const type = message.message_type || message.type || message.role;

  // Tool call messages — show tool name and arguments
  if (type === 'tool_call_message') {
    const calls = message.tool_calls || (message.tool_call ? [message.tool_call] : []);
    if (calls.length > 0) {
      return calls.map((tc: any) => {
        const args = tc.arguments || '';
        const truncated = args.length > 300 ? args.substring(0, 300) + '...' : args;
        return `${tc.name || 'unknown'}(${truncated})`;
      }).join('\n');
    }
    return null;
  }

  // Tool return messages — show status and return value
  if (type === 'tool_return_message') {
    const returns = message.tool_returns || [{
      status: message.status,
      tool_return: message.tool_return,
    }];
    if (returns.length > 0) {
      return returns.map((tr: any) => {
        const status = tr.status === 'error' ? '[ERROR] ' : '';
        const value = tr.tool_return || '';
        const truncated = value.length > 500 ? value.substring(0, 500) + '...' : value;
        return `${status}${truncated}`;
      }).join('\n');
    }
    return null;
  }

  // Reasoning messages — show reasoning content
  if (type === 'reasoning_message' && message.reasoning) {
    return message.reasoning;
  }

  // Existing fallback chain...
  if (message.text) return message.text;
  // ...
}
```

### Expected Result

```
10:58:10 PM [tool_call_message]
  start_image_generation({"company_id": "29bd7fec-...", "prompt": "Wide-angle view of a premium airplane cabin..."})

10:58:10 PM [tool_return_message]
  Image generation started. Job ID: aaa000b6-0fce-4dfa-a0ac-8905cb41b270
```

### Files to Change

| File | What |
|------|------|
| `src/commands/messages/utils.ts` | Add tool_call/tool_return/reasoning handling to `getMessageContent()` |

The display layer (`src/lib/ux/display/messages.ts`, `src/commands/runs/utils.ts`) already handles non-null content correctly — only the extraction needs fixing.

---

## Fix 2: Recover assistant response from false-failed runs

### Problem

When a Letta run is marked `failed`, the `send` command (async polling mode) immediately exits with code 1 and never attempts to extract the response:

```typescript
// src/commands/messages/send.ts lines 186-191
if (effectiveStatus === 'failed') {
  spinner.fail(`Message failed after ${timeStr}`);
  if (run.stop_reason) {
    error(`Reason: ${run.stop_reason}`);
  }
  process.exit(1);  // ← never checks for messages
}
```

### Evidence This Is Not an Edge Case

Full investigation of `run-ccdc7afc-2b90-4024-8e4b-5458fc831dc8`:

| Aspect | Value |
|--------|-------|
| Run status | `failed` |
| Run stop_reason | `error` |
| Run metadata.error | `""` (empty string) |
| Total steps | 14 |
| Steps with errors | **0** (every step has `error_type: None`, `error_data: None`) |
| Last step stop_reason | `end_turn` (success) |
| Last step prompt_tokens | 60,111 / 64,000 context limit (94% utilization) |
| Total messages | 42 |
| Tool calls with errors | **0** (all `status: success`, `is_err: null`) |
| assistant_message produced | **Yes** — correct IMAGE/VIDEO/CONCEPT output, last message in run |

The run completed all its work perfectly. The `failed` status was set by Letta server during post-run processing (likely context compaction at near-full context window). This will happen regularly with long agent conversations that approach the context limit.

### What Already Exists

`extractResponseText()` in `src/lib/messaging/bulk-messenger.ts` (lines 231-255) already knows how to extract assistant messages from a run — but it's private to the bulk-messenger and only called for completed runs.

`displayRunMessages()` in `src/commands/messages/send.ts` (lines 259-307) also extracts and displays assistant messages — but is only called on the `completed` path.

### Proposed Fix

In the `send` command's async polling loop, when a run fails, attempt to extract and display the response before reporting the failure:

```typescript
// src/commands/messages/send.ts — async polling failure handler
if (effectiveStatus === 'failed') {
  // Attempt to recover — Letta sometimes marks runs as failed
  // after the agent already produced its response (e.g. context
  // window pressure during post-run bookkeeping).
  const recovered = await tryRecoverResponse(client, runId);
  if (recovered) {
    spinner.warn(`Response from ${agent.name} (${timeStr}) [run reported failure but response was produced]:`);
    await displayRunMessages(client, runId, verbose, options.output);
    return;
  }

  spinner.fail(`Message failed after ${timeStr}`);
  if (run.stop_reason) {
    error(`Reason: ${run.stop_reason}`);
  }
  process.exit(1);
}
```

Where `tryRecoverResponse` checks if assistant messages exist:

```typescript
async function tryRecoverResponse(client: LettaClientWrapper, runId: string): Promise<boolean> {
  try {
    const messagesResponse = await client.getRunMessages(runId);
    const messages = normalizeResponse(messagesResponse);
    return messages.some((msg: any) =>
      msg.message_type === 'assistant_message' ||
      msg.type === 'assistant_message'
    );
  } catch {
    return false;
  }
}
```

This uses `spinner.warn()` (yellow) instead of `spinner.succeed()` (green) to signal the degraded state. The user sees the response but knows the run had issues.

### Same Fix in SDK's `waitForRun`

The SDK method `waitForRun()` in `src/sdk.ts` should also benefit — callers using lettactl programmatically (like adspectre's `async-client.ts`) need the same recovery. Consider adding a `recoverResponse` option or a richer return type that includes both status and any recovered response.

### Same Fix in Bulk Messenger

`src/lib/messaging/bulk-messenger.ts` — the `processAgent` function at the failure path (around line 125) should also attempt recovery and populate `result.responseText` when `collectResponse` is enabled, even for failed runs.

### Files to Change

| File | What |
|------|------|
| `src/commands/messages/send.ts` | Add recovery attempt before `process.exit(1)` in the failed handler |
| `src/lib/messaging/bulk-messenger.ts` | Attempt `extractResponseText()` for failed runs when `collectResponse` is set |
| `src/sdk.ts` | Consider enriching `waitForRun` return or adding recovery option |

---

## Summary

| Fix | Impact | Complexity |
|-----|--------|-----------|
| 1. Tool message display | DX improvement — no more hitting the API directly to debug agent tool calls | Low — single function change |
| 2. Failed run recovery | Correctness — prevents data loss when Letta server falsely marks successful runs as failed | Medium — 3 files, but the extraction logic already exists |

--- a/src/commands/messages/send.ts
+++ b/src/commands/messages/send.ts
@@ -4,11 +4,40 @@ import { normalizeResponse, sleep } from '../../lib/shared/response-normalizer';
 import { createSpinner, getSpinnerEnabled } from '../../lib/ux/spinner';
 import { sendMessageToAgent } from '../../lib/messaging/message-sender';
 import { bulkSendMessage } from '../../lib/messaging/bulk-messenger';
-import { log, output, error } from '../../lib/shared/logger';
+import { log, output, error, warn } from '../../lib/shared/logger';
 import { Run } from '../../types/run';
 import { SendOptions } from './types';
 import { getMessageContent, formatElapsedTime } from './utils';
 import { isRunTerminal, getEffectiveRunStatus } from '../../lib/messaging/run-utils';
+
+/**
+ * Attempt to recover assistant response from a run that was marked as failed.
+ * Returns true if a response was found and displayed, false otherwise.
+ */
+async function tryRecoverResponse(
+  client: LettaClientWrapper,
+  runId: string,
+  verbose: boolean,
+  outputFormat?: string,
+): Promise<boolean> {
+  try {
+    const messagesResponse = await client.getRunMessages(runId);
+    const messages = normalizeResponse(messagesResponse);
+
+    const hasAssistant = messages.some((msg: any) =>
+      msg.message_type === 'assistant_message' ||
+      msg.type === 'assistant_message' ||
+      (msg.role === 'assistant' && !msg.type?.includes('system'))
+    );
+
+    if (hasAssistant) {
+      return true;
+    }
+  } catch {
+    // Recovery is best-effort; fall through to normal failure path
+  }
+  return false;
+}
 
 export async function sendMessageCommand(
   agentNameOrMessage: string,
@@ -184,6 +213,12 @@ export async function sendMessageCommand(
         }
 
         if (effectiveStatus === 'failed') {
+          const recovered = await tryRecoverResponse(client, runId, verbose, options.output);
+          if (recovered) {
+            spinner.warn(`Run failed after ${timeStr} but response was recovered${run.stop_reason ? ` (${run.stop_reason})` : ''}`);
+            await displayRunMessages(client, runId, verbose, options.output);
+            return;
+          }
           spinner.fail(`Message failed after ${timeStr}`);
           if (run.stop_reason) {
             error(`Reason: ${run.stop_reason}`);

--- a/src/commands/messages/utils.ts
+++ b/src/commands/messages/utils.ts
@@ -2,6 +2,34 @@
  * Safely extracts content from different message types
  */
 export function getMessageContent(message: any): string | null {
+  const msgType = message.message_type || message.type;
+
+  // Tool call messages: show "toolName(args)"
+  if (msgType === 'tool_call_message') {
+    const name = message.tool_call?.name || message.name || 'unknown_tool';
+    let args = '';
+    const rawArgs = message.tool_call?.arguments || message.arguments;
+    if (rawArgs) {
+      args = typeof rawArgs === 'string' ? rawArgs : JSON.stringify(rawArgs);
+      if (args.length > 300) args = args.slice(0, 300) + '...';
+    }
+    return `${name}(${args})`;
+  }
+
+  // Tool return messages: show return value (prefixed with [ERROR] if status is error)
+  if (msgType === 'tool_return_message') {
+    let value = message.tool_return || message.return_value || message.content || '';
+    if (typeof value !== 'string') value = JSON.stringify(value);
+    if (value.length > 500) value = value.slice(0, 500) + '...';
+    const isError = message.status === 'error';
+    return isError ? `[ERROR] ${value}` : value;
+  }
+
+  // Reasoning messages
+  if (msgType === 'reasoning_message') {
+    return message.reasoning || null;
+  }
+
   // Try different properties that might contain the message content
   if (message.text) return message.text;
   if (message.content) {

--- a/src/lib/messaging/bulk-messenger.ts
+++ b/src/lib/messaging/bulk-messenger.ts
@@ -125,6 +125,9 @@ export async function bulkSendMessage(
           } else if (effectiveStatus === 'failed') {
             result.status = 'failed';
             result.error = runStatus.stop_reason ? `Run failed: ${runStatus.stop_reason}` : 'Run failed';
+            if (options.collectResponse) {
+              result.responseText = await extractResponseText(client, run.id);
+            }
             break;
           } else if (effectiveStatus === 'cancelled') {
             result.status = 'cancelled';
@@ -153,6 +156,8 @@ export async function bulkSendMessage(
     const durationStr = result.duration.toFixed(1);
     if (result.status === 'completed') {
       outputFn(`OK ${agent.name} (${durationStr}s)`);
+    } else if (result.status === 'failed' && result.responseText) {
+      outputFn(`WARN ${agent.name}: ${result.error || 'unknown error'} (response recovered)`);
     } else {
       outputFn(`FAIL ${agent.name}: ${result.error || 'unknown error'}`);
     }

--- a/src/lib/ux/spinner.ts
+++ b/src/lib/ux/spinner.ts
@@ -5,6 +5,7 @@ export interface SpinnerInterface {
   text: string;
   start(): SpinnerInterface;
   succeed(text?: string): SpinnerInterface;
+  warn(text?: string): SpinnerInterface;
   fail(text?: string): SpinnerInterface;
   stop(): SpinnerInterface;
 }
@@ -18,6 +19,11 @@ class NoSpinner implements SpinnerInterface {
 
   succeed(text?: string): SpinnerInterface {
     if (text) output(`[OK] ${text}`);
+    return this;
+  }
+
+  warn(text?: string): SpinnerInterface {
+    if (text) output(`[WARN] ${text}`);
     return this;
   }
 
@@ -39,6 +45,10 @@ class QuietSpinner implements SpinnerInterface {
   }
 
   succeed(_text?: string): SpinnerInterface {
+    return this;
+  }
+
+  warn(_text?: string): SpinnerInterface {
     return this;
   }
 

--- a/tests/unit/commands/get-message-content.test.ts
+++ b/tests/unit/commands/get-message-content.test.ts
@@ -1,0 +1,163 @@
+import { getMessageContent } from '../../../src/commands/messages/utils';
+
+describe('getMessageContent', () => {
+  describe('tool_call_message', () => {
+    it('should display tool name and arguments', () => {
+      const msg = {
+        message_type: 'tool_call_message',
+        tool_call: { name: 'archival_memory_search', arguments: '{"query":"hello"}' },
+      };
+      expect(getMessageContent(msg)).toBe('archival_memory_search({"query":"hello"})');
+    });
+
+    it('should handle object arguments', () => {
+      const msg = {
+        message_type: 'tool_call_message',
+        tool_call: { name: 'send_message', arguments: { message: 'hi' } },
+      };
+      expect(getMessageContent(msg)).toBe('send_message({"message":"hi"})');
+    });
+
+    it('should truncate arguments at 300 chars', () => {
+      const longArgs = 'x'.repeat(400);
+      const msg = {
+        message_type: 'tool_call_message',
+        tool_call: { name: 'my_tool', arguments: longArgs },
+      };
+      const result = getMessageContent(msg)!;
+      expect(result).toMatch(/^my_tool\(/);
+      expect(result).toMatch(/\.\.\.\)$/);
+      // "my_tool(" = 8, truncated args = 300, "..." = 3, ")" = 1
+      expect(result.length).toBe(8 + 300 + 3 + 1);
+    });
+
+    it('should show empty parens when no arguments', () => {
+      const msg = {
+        message_type: 'tool_call_message',
+        tool_call: { name: 'heartbeat' },
+      };
+      expect(getMessageContent(msg)).toBe('heartbeat()');
+    });
+
+    it('should fall back to message.name when tool_call.name missing', () => {
+      const msg = {
+        message_type: 'tool_call_message',
+        name: 'fallback_tool',
+        arguments: '{}',
+      };
+      expect(getMessageContent(msg)).toBe('fallback_tool({})');
+    });
+
+    it('should show unknown_tool when no name available', () => {
+      const msg = { message_type: 'tool_call_message' };
+      expect(getMessageContent(msg)).toBe('unknown_tool()');
+    });
+
+    it('should work with type instead of message_type', () => {
+      const msg = {
+        type: 'tool_call_message',
+        tool_call: { name: 'search', arguments: '{"q":"test"}' },
+      };
+      expect(getMessageContent(msg)).toBe('search({"q":"test"})');
+    });
+  });
+
+  describe('tool_return_message', () => {
+    it('should display return value', () => {
+      const msg = {
+        message_type: 'tool_return_message',
+        tool_return: 'Search returned 3 results',
+      };
+      expect(getMessageContent(msg)).toBe('Search returned 3 results');
+    });
+
+    it('should prefix with [ERROR] when status is error', () => {
+      const msg = {
+        message_type: 'tool_return_message',
+        tool_return: 'Connection refused',
+        status: 'error',
+      };
+      expect(getMessageContent(msg)).toBe('[ERROR] Connection refused');
+    });
+
+    it('should not prefix when status is not error', () => {
+      const msg = {
+        message_type: 'tool_return_message',
+        tool_return: 'OK',
+        status: 'success',
+      };
+      expect(getMessageContent(msg)).toBe('OK');
+    });
+
+    it('should truncate return value at 500 chars', () => {
+      const msg = {
+        message_type: 'tool_return_message',
+        tool_return: 'y'.repeat(600),
+      };
+      const result = getMessageContent(msg)!;
+      expect(result).toHaveLength(500 + 3); // 500 + "..."
+      expect(result).toMatch(/\.\.\.$/);
+    });
+
+    it('should handle object return values', () => {
+      const msg = {
+        message_type: 'tool_return_message',
+        tool_return: { results: [1, 2, 3] },
+      };
+      expect(getMessageContent(msg)).toBe('{"results":[1,2,3]}');
+    });
+
+    it('should fall back to return_value then content', () => {
+      expect(getMessageContent({
+        message_type: 'tool_return_message',
+        return_value: 'from return_value',
+      })).toBe('from return_value');
+
+      expect(getMessageContent({
+        message_type: 'tool_return_message',
+        content: 'from content',
+      })).toBe('from content');
+    });
+
+    it('should return empty string when no value available', () => {
+      const msg = { message_type: 'tool_return_message' };
+      expect(getMessageContent(msg)).toBe('');
+    });
+  });
+
+  describe('reasoning_message', () => {
+    it('should return reasoning text', () => {
+      const msg = {
+        message_type: 'reasoning_message',
+        reasoning: 'The user wants to search their memory for past conversations.',
+      };
+      expect(getMessageContent(msg)).toBe(
+        'The user wants to search their memory for past conversations.'
+      );
+    });
+
+    it('should return null when reasoning is empty', () => {
+      const msg = { message_type: 'reasoning_message' };
+      expect(getMessageContent(msg)).toBeNull();
+    });
+  });
+
+  describe('existing message types (unchanged)', () => {
+    it('should return text property', () => {
+      expect(getMessageContent({ text: 'hello' })).toBe('hello');
+    });
+
+    it('should return string content', () => {
+      expect(getMessageContent({ content: 'world' })).toBe('world');
+    });
+
+    it('should handle content arrays', () => {
+      const msg = { content: [{ text: 'part1' }, { content: 'part2' }] };
+      expect(getMessageContent(msg)).toBe('part1 part2');
+    });
+
+    it('should return null for empty messages', () => {
+      expect(getMessageContent({})).toBeNull();
+    });
+  });
+});

--- a/tests/unit/commands/messages.test.ts
+++ b/tests/unit/commands/messages.test.ts
@@ -9,8 +9,10 @@ jest.mock('ora', () => {
   return jest.fn(() => ({
     start: jest.fn(() => ({
       succeed: jest.fn(),
+      warn: jest.fn(),
       fail: jest.fn(),
       stop: jest.fn(),
+      text: '',
     })),
   }));
 });
@@ -146,6 +148,74 @@ describe('messages commands', () => {
         messages: [{ role: 'user', content: 'Hello' }],
         streaming: true
       });
+    });
+
+    it('should recover response from a false-failed run', async () => {
+      const mockAgent = { id: 'agent-123', name: 'test-agent' };
+      const mockRunResponse = { id: 'run-789', status: 'pending' };
+
+      mockResolver.findAgentByName.mockResolvedValue({ agent: mockAgent, allAgents: [] });
+      mockClient.createAsyncMessage.mockResolvedValue(mockRunResponse as any);
+
+      // First poll returns failed run
+      mockClient.getRun.mockResolvedValue({
+        id: 'run-789',
+        status: 'failed',
+        stop_reason: 'context_window_overflow_in_system_prompt',
+      } as any);
+
+      // getRunMessages returns messages including an assistant_message
+      mockClient.getRunMessages.mockResolvedValue([
+        { message_type: 'reasoning_message', reasoning: 'thinking...' },
+        { message_type: 'assistant_message', content: 'Here is the recovered response' },
+      ] as any);
+
+      await sendMessageCommand('test-agent', 'Hello agent', {}, mockCommand);
+
+      // Should have called getRunMessages to attempt recovery
+      expect(mockClient.getRunMessages).toHaveBeenCalledWith('run-789');
+
+      // Should output the recovered response (not exit with failure)
+      const allOutputs = mockConsoleLog.mock.calls.map(c => c[0]);
+      const hasRecoveredContent = allOutputs.some(
+        (msg: string) => typeof msg === 'string' && msg.includes('Here is the recovered response')
+      );
+      expect(hasRecoveredContent).toBe(true);
+    });
+
+    it('should exit on failed run with no assistant response to recover', async () => {
+      const mockAgent = { id: 'agent-123', name: 'test-agent' };
+      const mockRunResponse = { id: 'run-999', status: 'pending' };
+
+      mockResolver.findAgentByName.mockResolvedValue({ agent: mockAgent, allAgents: [] });
+      mockClient.createAsyncMessage.mockResolvedValue(mockRunResponse as any);
+
+      mockClient.getRun.mockResolvedValue({
+        id: 'run-999',
+        status: 'failed',
+        stop_reason: 'error',
+      } as any);
+
+      // getRunMessages returns only non-assistant messages
+      mockClient.getRunMessages.mockResolvedValue([
+        { message_type: 'reasoning_message', reasoning: 'thinking...' },
+        { message_type: 'tool_call_message', tool_call: { name: 'search', arguments: '{}' } },
+      ] as any);
+
+      const mockExit = jest.spyOn(process, 'exit').mockImplementation((() => {
+        throw new Error('process.exit called');
+      }) as any);
+
+      try {
+        await sendMessageCommand('test-agent', 'Hello agent', {}, mockCommand);
+        // Should not reach here
+        expect(true).toBe(false);
+      } catch (e: any) {
+        expect(e.message).toBe('process.exit called');
+      }
+
+      expect(mockExit).toHaveBeenCalledWith(1);
+      mockExit.mockRestore();
     });
   });
 });

